rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp

```diff
@@ -136,60 +151,58 @@ class rmw_context_impl_s::Data final
-      call_success = z_call(channel.recv, &reply))
-    {
-      if (!call_success) {
-        continue;
-      }
-      if (z_reply_is_ok(&reply)) {
-        z_sample_t sample = z_reply_ok(&reply);
-        z_owned_str_t keystr = z_keyexpr_to_string(sample.keyexpr);
+      z_loan(session_), z_loan(keyexpr),
+      z_move(closure), NULL);
+
+    z_owned_reply_t reply;
+    while (z_recv(z_loan(handler), &reply) == Z_OK) {
+      if (z_reply_is_ok(z_loan(reply))) {
+        const z_loaned_sample_t * sample = z_reply_ok(z_loan(reply));
+        z_view_string_t keystr;
+        z_keyexpr_as_view_string(z_sample_keyexpr(sample), &keystr);
+        std::string str(z_string_data(z_loan(keystr)), z_string_len(z_loan(keystr)));
         // Ignore tokens from the same session to avoid race conditions from this
         // query and the liveliness subscription.
-        graph_cache_->parse_put(z_loan(keystr), true);
-        z_drop(z_move(keystr));
+        graph_cache_->parse_put(str, true);
       } else {
         RMW_ZENOH_LOG_DEBUG_NAMED(
           "rmw_zenoh_cpp", "[rmw_context_impl_s] z_call received an invalid reply.\n");
       }
+      z_drop(z_move(reply));
     }
-    z_drop(z_move(reply));
-    z_drop(z_move(channel));
+    z_drop(z_move(handler));
 
     // Initialize the shm manager if shared_memory is enabled in the config.
-    shm_manager_ = std::nullopt;
-    if (shm_enabled._cstr != nullptr &&
-      strcmp(shm_enabled._cstr, "true") == 0)
+    shm_provider_ = std::nullopt;
+    if (strncmp(
+        z_string_data(z_loan(shm_enabled)),
+        "true",
+        z_string_len(z_loan(shm_enabled))) == 0)
     {
-      char idstr[sizeof(zid.id) * 2 + 1];  // 2 bytes for each byte of the id, plus the trailing \0
-      static constexpr size_t max_size_of_each = 3;  // 2 for each byte, plus the trailing \0
-      for (size_t i = 0; i < sizeof(zid.id); ++i) {
-        snprintf(idstr + 2 * i, max_size_of_each, "%02x", zid.id[i]);
-      }
-      idstr[sizeof(zid.id) * 2] = '\0';
-      // TODO(yadunund): Can we get the size of the shm from the config even though it's not
-      // a standard parameter?
-      shm_manager_ =
-        zc_shm_manager_new(
-        z_loan(session_),
-        idstr,
-        SHM_BUFFER_SIZE_MB * 1024 * 1024);
-      if (!shm_manager_.has_value() ||
-        !zc_shm_manager_check(&shm_manager_.value()))
-      {
-        throw std::runtime_error("Unable to create shm manager.");
+      // TODO(yuyuan): determine the default alignment of SHM
+      z_alloc_alignment_t alignment = {5};
+      z_owned_memory_layout_t layout;
+      z_memory_layout_new(&layout, SHM_BUFFER_SIZE_MB * 1024 * 1024, alignment);
+
+      z_owned_shm_provider_t provider;
+      if (z_posix_shm_provider_new(&provider, z_loan(layout)) != Z_OK) {
+        RMW_ZENOH_LOG_ERROR_NAMED("rmw_zenoh_cpp", "Unable to create a SHM provider.");
+        throw std::runtime_error("Unable to create shm provider.");
       }
+      shm_provider_ = provider;
```

Comment:
Does this need to be a z_move?

Reply:
This doesn't work in the sense of `z_move`. `z_move` converts an owned object into a moved object. And it will be automatically dropped after another usage of zenoh functions. After we changed to zenoh-cpp, `std::move` is more appropriate to use.

To further clarify the zenoh-c mechanism, 
```c
z_owned_bytes_t a;
z_owned_bytes_t b;
b = a;                 // This duplicates the pointer to the zenoh data.
z_bytes_clone(&b, a);  // This shallow copies the underly data by incrementing the reference counter.
```